### PR TITLE
fix: [DHIS2-9286] add exception handling globally to catch any unhandled ones and notify job correctly

### PR DESCRIPTION
**Pull request by ameenhere**
Added an try catch block around the new tracker import service flow, to catch any unhandled exceptions. 
In case of exceptions, the import report and job summary in notifier will reflect the same with exception error message in it.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/DefaultTrackerImportService.java [hunk 1]
```diff
@@ -102,32 +104,62 @@ public TrackerImportReport importTracker( TrackerImportParams params )
...
+            importReport.setStatus( TrackerStatus.ERROR );
+            
+            if ( params.hasJobConfiguration() )
+            {
+                notifier.update( params.getJobConfiguration(), "(" + params.getUsername() + ") Import:Failed with exception: " + e.getMessage(), true );
+                notifier.addJobSummary( params.getJobConfiguration(), importReport, TrackerImportReport.class );
+            }
+            
+            //TODO: Should this be rethrown in case of sync? Or is the error report added to the import report in the above lines enough?
+            throw e;
```
Comment (ameenhere) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/DefaultTrackerImportService.java:159]: I think we shouldnt rethrow this exception. Your thoughts?

Comment (luciano-fiandesio) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/DefaultTrackerImportService.java]: I don't think we should rethrow. If we rethrow, then the exception will be "handled" by the `SecurityContextRunnable` class (or a subclass). Please, see my comment in the approve box, based on our conversation.

**Review by luciano-fiandesio — APPROVED** [review]
Leaving here a consideration, based on a discussion we had with Ameen.
This comment is focused on the async flow since the sync flow it is not implemented.
The fix in this PR is perfectly fine because it tries to catch *every* exception that can happen during an import flow.
If we would not catch the exceptions at this "level", an exception would bubble up to the `SecurityContextRunnable` class (or a subclass). This class is part of the job management framework. This class has a `handleError` method, which should be implemented in the subclasses in order to handle exceptions. The problem here is that, in that method, we have no visibility of the Tracker Report class.
So, from an architectural standpoint, the solution in this PR is not 100% correct, because it does not use the underlying framework which has been designed to handle exceptions in the async thread. 
@stian-sandvold @mortenoh thoughts?

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/DefaultTrackerImportService.java [hunk 2]
```diff
@@ -102,32 +107,60 @@ public TrackerImportReport importTracker( TrackerImportParams params )
...
+            log.error( "Exception thrown during import.",e );
+            
+            TrackerValidationReport tvr = importReport.getTrackerValidationReport();
+            
+            if ( tvr == null )
+            {
+                tvr = new TrackerValidationReport();
+            }
+            
+            tvr.getErrorReports().add(new TrackerErrorReport( null, "Exception:"+e.getMessage(), TrackerErrorCode.E9999, 0, null, null, null, null ));
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/DefaultTrackerImportService.java:152]: There is a static method to create new errors ValidationErrorReporter.newReport(errorCode) and then you can add an arg.
You can find a lot of examples in the validation hooks.
Not a big change, but we avoid to have all this null in the constructor

**Review by enricocolasante — APPROVED** [review]
Just one small comment

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/DefaultTrackerImportService.java (hunk 2 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/DefaultTrackerImportService.java:152]: Good point! Thanks

**Pull request merged**